### PR TITLE
[SYCL] Use mp11 to sort compile-time properties

### PR DESCRIPTION
**Pull request by uditagarwal97**
Currently, compile-time properties are sorted using a hand-written implementation of merge-sort. However, this PR propose to sort compile-time properties using Boost's MP11 metaprogramming library instead. Mp11's sorting functionality is more optimized and it also improves the overall code readability (83 LOC vs just 11 LOC).

PS:- This PR does not introduce any new dependency as Mp11 is already integrated into the code base (See [#5791](https://github.com/intel/llvm/pull/5791)).

### CODE REVIEW

File: sycl/include/sycl/ext/oneapi/properties/property_utils.hpp [hunk 1]
```diff
@@ -10,9 +10,12 @@
 
 #include <sycl/detail/property_helper.hpp>
 #include <sycl/ext/oneapi/properties/property.hpp>
+#include <sycl/detail/boost/mp11.hpp>
 
 #include <tuple>
 
+// Using mp11 to sort property lists.
+namespace __MP11_NS = sycl::detail::boost::mp11;
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/properties/property_utils.hpp:18]: I think it's better not to introduce that in the global namespace.

File: sycl/include/sycl/ext/oneapi/properties/property_utils.hpp [hunk 2]
```diff
@@ -106,93 +109,17 @@ template <typename RHS> struct SelectNonVoid<void, RHS> {
...
-// Merges two tuples by recursively extracting the type with the minimum
-// PropertyID in the two tuples and prepending it to the merging of the
-// remaining elements.
-template <typename T1, typename T2> struct Merge {};
-template <typename... LTs> struct Merge<std::tuple<LTs...>, std::tuple<>> {
-  using type = std::tuple<LTs...>;
+// Sort types accoring to their PropertyID.
+struct SortByPropertyId {
+  template<typename T, typename U> using fn =
+      __MP11_NS::mp_bool<PropertyID<T>::value < PropertyID<U>::value>;
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/properties/property_utils.hpp:115]: ... i.e, use fully qualified name here,

File: sycl/include/sycl/ext/oneapi/properties/property_utils.hpp [hunk 3]
```diff
@@ -106,93 +109,17 @@ template <typename RHS> struct SelectNonVoid<void, RHS> {
...
-  using type = typename MergeAll<reduced>::type;
-};
-
-// Performs merge-sort on types with PropertyID.
 template <typename... Ts> struct Sorted {
   static_assert(detail::AllPropertyValues<std::tuple<Ts...>>::value,
                 "Unrecognized property in property list.");
-  using split = typename CreateTuplePairs<Ts...>::type;
-  using type = typename MergeAll<split>::type;
+  using properties = __MP11_NS::mp_list<Ts...>;
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/properties/property_utils.hpp:120]: ... can use type alias or `using namespace` declaration in this scope though.

File: sycl/include/sycl/ext/oneapi/properties/property_utils.hpp [hunk 4]
```diff
@@ -106,93 +109,17 @@ template <typename RHS> struct SelectNonVoid<void, RHS> {
...
 
-// Merges two tuples by recursively extracting the type with the minimum
-// PropertyID in the two tuples and prepending it to the merging of the
-// remaining elements.
-template <typename T1, typename T2> struct Merge {};
-template <typename... LTs> struct Merge<std::tuple<LTs...>, std::tuple<>> {
-  using type = std::tuple<LTs...>;
+// Sort types accoring to their PropertyID.
+struct SortByPropertyId {
+  template<typename T, typename U> using fn =
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/properties/property_utils.hpp:114]: Can we use the same T1/T2 as on the left? I'm still not sure what our policies regarding those identifiers are.

**Review by aelovikov-intel — COMMENTED** [review]
LGTM, the tests changes are due to the change in dumps that are non-functional, similar to what can be seen at https://godbolt.org/z/ndan4djdb

@steffenlarsen , do you agree with this? An alternative would be to update our boost machinery to use `sycl::_V1::detail::boost` instead of current `sycl::detail::boost`.

**Comment by steffenlarsen** [comment]
> @steffenlarsen , do you agree with this? An alternative would be to update our boost machinery to use `sycl::_V1::detail::boost` instead of current `sycl::detail::boost`.

I am okay with either, though I prefer the one that's most unlikely to cause the need to change the test in the future. Would the latter localize it to the actual changes?

Note also that RHEL build is failing, seemingly due some new ambiguities in unchanged code. First issue I see is use of `sycl::detail::conditional_t`, which should now be fixable by just changing it to use `std::conditional_t`.

**Comment by uditagarwal97** [comment]
Thanks for the feedback @steffenlarsen  and @aelovikov-intel . Instead of my previous changes in the test cases, I have now made changes to import boost into sycl::_V1::detail namespace instead of the sycl::detail namespace. This prevents the need for updating test cases and also fixes the errors in pre-commit testing ([jenkins / SYCL_CI/intel/Build_PR_RHEL / Build_PR_RHEL / #18327](http://icl-jenkins.sc.intel.com:8080/blue/organizations/jenkins/SYCL_CI%2Fintel%2FBuild_PR_RHEL/detail/Build_PR_RHEL/18327/pipeline/)).

File: sycl/include/sycl/ext/oneapi/properties/property_utils.hpp [hunk 5]
```diff
@@ -106,93 +107,20 @@ template <typename RHS> struct SelectNonVoid<void, RHS> {
...
-// PropertyID in the two tuples and prepending it to the merging of the
-// remaining elements.
-template <typename T1, typename T2> struct Merge {};
-template <typename... LTs> struct Merge<std::tuple<LTs...>, std::tuple<>> {
-  using type = std::tuple<LTs...>;
+// Sort types accoring to their PropertyID.
+struct SortByPropertyId {
+  template <typename T1, typename T2>
+  using fn = sycl::detail::boost::mp11::mp_bool <
+             PropertyID<T1>::value<PropertyID<T2>::value>;
```
Comment (steffenlarsen) [inline sycl/include/sycl/ext/oneapi/properties/property_utils.hpp:114]: Nit; Just to make it a little easier to read:

```suggestion
             (PropertyID<T1>::value < PropertyID<T2>::value)>;
```
Hopefully the formatter agrees with this.

**Review by steffenlarsen — APPROVED** [review]
Sad to see it go, but thank you for taking care of this! LGTM with minor nit. 🚀

**Review by aelovikov-intel — APPROVED** [review]

**Comment by uditagarwal97** [comment]
@llvm-gatekeepers The PR is ready!

**Pull request merged**